test/BatchExchange.spec.ts

```diff
@@ -0,0 +1,3000 @@
+        0,
+        "Second order should be removed",
+      );
+      assert.equal(
+        toOrderResult(
+          await batchExchange.orders(user_1, 2),
+        ).sellToken.toNumber(),
+        6,
+        "Third order should be present",
+      );
+      assert.equal(
+        toOrderResult(
+          await batchExchange.orders(user_1, 3),
+        ).sellToken.toNumber(),
+        7,
+        "Fourth order should be present",
+      );
+    });
+  });
+  describe("submitSolution()", () => {
+    it("rejects if claimed objective is not better than current", async () => {
+      const batchExchange = await setupGenericStableX();
+
+      const batchId = (await batchExchange.getCurrentBatchId()).toNumber();
+      await closeAuction(batchExchange);
+
+      await truffleAssert.reverts(
+        batchExchange.submitSolution(batchId, 0, [], [], [], [], []),
+        "Claimed objective doesn't sufficiently improve current solution",
+      );
+    });
+    it("rejects trivial solution (the only solution with zero utility)", async () => {
+      const batchExchange = await setupGenericStableX();
+      const batchId = (await batchExchange.getCurrentBatchId()).toNumber();
+      await closeAuction(batchExchange);
+      const fakeClaimedObjective = 1;
+      await truffleAssert.reverts(
+        batchExchange.submitSolution(
+          batchId,
+          fakeClaimedObjective,
+          [],
+          [],
+          [],
+          [toETH(1)],
+          [1],
+          { from: solver },
+        ),
+        "New objective doesn't sufficiently improve current solution",
+      );
+    });
+    it("[Basic Trade] places two orders and returns calculated utility", async () => {
+      const batchExchange = await setupGenericStableX();
+
+      // Make deposits, place orders and close auction[aka runAuctionScenario(basicTrade)]
+      await makeDeposits(batchExchange, accounts, basicTrade.deposits);
+      const batchId = (await batchExchange.getCurrentBatchId()).toNumber();
+      const orderIds = await placeOrders(
+        batchExchange,
+        accounts,
+        basicTrade.orders,
+        batchId + 1,
+      );
+      await closeAuction(batchExchange);
+
+      const solution = solutionSubmissionParams(
+        basicTrade.solutions[0],
+        accounts,
+        orderIds,
+      );
+
+      // Note: the claimed objective value is intentionally incorrect, this is to make sure that a call
+      // to `submitSolution` can be used to acurately determine the objective value of a solution
+      const objectiveValue = await batchExchange.submitSolution.call(
+        batchId,
+        1,
+        solution.owners,
+        solution.touchedorderIds,
+        solution.volumes,
+        solution.prices,
+        solution.tokenIdsForPrice,
+        { from: solver },
+      );
+
+      assert(
+        objectiveValue.gtn(0),
+        "the computed objective value is greater than 0",
+      );
+      assert(objectiveValue.eq(solution.objectiveValue));
```

Comment:
Including both assertions for context:

The first assertion here does not seem necessary as it is trumped by the second. If we do chose to keep it, however, I think the failure string should be modified to its negation?